python/GafferTest/AnimationTest.py

```diff
@@ -407,6 +407,53 @@ def testUndoRemoveKey( self ) :
 		self.assertFalse( curve.hasKey( key.getTime() ) )
 		self.assertEqual( key.parent(), None )
 
+	def testUndoAddRemoveKeyWithInterveningTimeChange( self ) :
+
+		s = Gaffer.ScriptNode()
+
+		s["n"] = Gaffer.Node()
+		s["n"]["user"]["f"] = Gaffer.FloatPlug( flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )
+
+		curve = Gaffer.Animation.acquire( s["n"]["user"]["f"] )
+		self.assertFalse( curve.getKey( 0 ) )
+
+		key1 = Gaffer.Animation.Key( 1, 0 )
+
+		with Gaffer.UndoScope( s ) :
+			curve.addKey( key1 )
+			self.assertEqual( curve.getKey( 1 ), key1 )
+
+		with Gaffer.UndoScope( s ) :
+			curve.removeKey( key1 )
+			self.assertIsNone( curve.getKey( 1 ) )
+
+		# NOTE : this operation is not captured by the undo system as key is not in curve
+		with Gaffer.UndoScope( s ) :
+			key1.setTime( 2 )
+			self.assertEqual( key1.getTime(), 2 )
+
+		with Gaffer.UndoScope( s ) :
+			curve.addKey( key1 )
+			self.assertEqual( curve.getKey( 2 ), key1 )
+
+		s.undo()
+		self.assertIsNone( curve.getKey( 2 ) )
+
+		s.undo()
+		self.assertEqual( curve.getKey( 1 ), key1 )
```

Comment:
I think there's a bit of a debate to be had here, and I find myself wanting to make the case that this is an unreasonable expectation. `Key` was intended to roughly mimic the pattern established by `GraphComponent`, with `addKey()/removeKey()` being analogous to `addChild/removeChild()`, and `getTime()` being analogous to `getName()` (CurvePlug indexes by time, GraphComponent indexes by name). And the very same "bug" exists for GraphComponent, if you change the name of something while it's removed. But the equivalent "fix" for GraphComponent would be to restore its entire state, which is completely intractable (they have their own children, connections to plugs etc). I don't think we should encourage that expectation.

To put this in other words, `undo()` only reverts the effects of the specific action it is associated with. It is _not_ intended to also magically revert the effects of anything not captured by the undo system. If you want to be able to undo/redo back to the exact state you were in, then you need to perform all operations within the undo system.

Here's a unit test that demonstrates this perspective, and which passes currently. Note that `addChild()/removeChild()` (and their equivalent undo/redo) do still provide solid postconditions as expressed by `assertParented()` and `assertNotParented()`, they just don't factor in magical state tracking outside the undo system. Transposing that to `addKey()/removeKey()`, I believe the current master would pass too.

<details>

```
	def testUndoAddRemoveChildWithInterveningNameChange( self ) :

		s = Gaffer.ScriptNode()

		s["n"] = Gaffer.Node()
		parent = s["n"]["user"]
		child = Gaffer.Plug( "name1" )

		def assertParented() :

			self.assertEqual( child.parent(), parent )
			self.assertEqual( parent.getChild( child.getName() ), child )
			self.assertEqual( parent.keys(), [ child.getName() ] )
			self.assertEqual( parent.values(), [ child ] )

		def assertNotParented() :

			self.assertEqual( child.parent(), None )
			self.assertIsNone( parent.getChild( child.getName() ) )
			self.assertEqual( parent.keys(), [] )
			self.assertEqual( parent.values(), [] )

		assertNotParented()

		with Gaffer.UndoScope( s ) :
			parent.addChild( child )

		assertParented()

		with Gaffer.UndoScope( s ) :
			parent.removeChild( child )

		assertNotParented()

		child.setName( "name2" )
		self.assertEqual( child.getName(), "name2" )

		assertNotParented()

		with Gaffer.UndoScope( s ) :
			parent.addChild( child )

		assertParented()

		s.undo()
		assertNotParented()

		s.undo()
		assertParented()

		s.undo()
		assertNotParented()

		s.redo()
		assertParented()

		s.redo()
		assertNotParented()

		s.redo()
		assertParented()
```
</details>

Now, this is all well in _theory_, but obviously a practical problem led to this change, and we have to deal with reality. Am I right in thinking this was related to the AnimationEditor storing keys temporarily outside the curve when they get temporarily clobbered during a drag operation? What I'm wondering is if we can deal with that within the constraints I've outlined above. I guess the first but of reality I'm ignorant of is why we need to modify the keys when they're outside the undo system...

Reply:
I think I can imagine myself expecting the behaviour that Paul has implemented, just because a Node feels like a persistent object that I'm merely reparenting, and there will still be only one of, whereas a Key feels like something so light and simple that it could just be copied.

But if we wanted to support that mental model, then that would suggest going further and actually copying, so that it was supported to add 3 keys with:
```k = Gaffer.Animation.Key( 1, 0 )
curve.addKey( k )
k.setTime( 2 )
curve.addKey( k )
k.setTime( 3 )
curve.addKey( k )```

Reply:
> whereas a Key feels like something so light and simple that it could just be copied. But if we wanted to support that mental model, then that would suggest going further and actually copying...

In practice Key identity is actually pretty important - it's needed for managing selection in the AnimationEditor, and for widgets which display keys to know what they are displaying. So a "copying" approach would end up requiring a "persistent key handle" anyway, at which point you may as well have distinct keys as we do. What I'm shooting for here is _less_ change to the existing design rather than more, which is why I've raised the question about the change to the undo model.

@proberts-cinesite, I'm guessing we're modifying orphaned keys somewhere in the AnimationEditor, maybe while they're temporarily orphaned by a drag-move? Why do we end up doing that, and is it avoidable?

Reply:
As I mentioned the problem is that a key's time can be changed whilst the key is unparented, this is not captured by the undo, but the undo/redo of add/remove key relies on the keys being in the same place as they were, if they are not its not possible to undo/redo correctly, the existing code crashes in this case as the search returns a null pointer which is then dereferenced to set time ```key->m_time = ...```. So its currently possible to write python code that will crash. Is that considered a bug?

Reply:
> So its currently possible to write python code that will crash. Is that considered a bug?

Yes, that would definitely be considered a bug - we strive to ensure that Python code will never crash. There are a couple of spots where we haven't been able to make that absolutely watertight, but I think we should aim to here.

Here's a unit test I'm proposing for the expected behaviour for this - it mirrors the test I posted above for GraphComponent, and it doesn't crash currently. The new test in this PR doesn't crash either as far as I can tell (I removed the asserts and ran it on loop) unless I missed something?

<details>

```
	def testUndoAddRemoveKeyWithInterveningTimeChange( self ) :

		s = Gaffer.ScriptNode()

		s["n"] = Gaffer.Node()
		s["n"]["user"]["f"] = Gaffer.FloatPlug( flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic )

		curve = Gaffer.Animation.acquire( s["n"]["user"]["f"] )
		key = Gaffer.Animation.Key( 1, 0 )

		def assertParented() :

			self.assertEqual( key.parent(), curve )
			self.assertEqual( curve.getKey( key.getTime() ), key )
			self.assertIsNone( curve.nextKey( key.getTime() ) )
			self.assertIsNone( curve.previousKey( key.getTime() ) )

		def assertNotParented() :

			self.assertEqual( key.parent(), None )
			self.assertIsNone( curve.getKey( key.getTime() ) )
			self.assertIsNone( curve.nextKey( key.getTime() ) )
			self.assertIsNone( curve.previousKey( key.getTime() ) )

		assertNotParented()

		with Gaffer.UndoScope( s ) :
			curve.addKey( key )

		assertParented()

		with Gaffer.UndoScope( s ) :
			curve.removeKey( key )

		assertNotParented()

		key.setTime( 2 )
		self.assertEqual( key.getTime(), 2 )

		assertNotParented()

		with Gaffer.UndoScope( s ) :
			curve.addKey( key )

		assertParented()

		s.undo()
		assertNotParented()

		s.undo()
		assertParented()

		s.undo()
		assertNotParented()

		s.redo()
		assertParented()

		s.redo()
		assertNotParented()

		s.redo()
		assertParented()
```
</details>

Reply:
As discussed, the code will now store the expected time of a key and throw an exception if the key's time is not as expected during undo/redo. This maintains the principle outlined above and gives feedback to the api user regarding incorrect api usage.

Reply:
As also discussed, setting the time of multiple keys where the new times would clash with existing keys is a complex operation to implement correctly using the current api, so we will add new api functions to make this easier and implement as a single logical operation where keys logically do not have there time changed whilst unparented, this will be implemented in a separate pr.

Reply:
:+1: For anyone following along, the other thing we just discussed was the reason this is an issue for the Animation API in particular. This is down to the need to move a bunch of keys in time during drags in the AnimationEditor. Moving a key can temporarily clobber other keys, which need to be unparented and then possibly reparented later.  The obvious `for key in selection : key.setTime( key.getTime() + 1 )` just doesn't work, and getting it right requires a non-trivial amount of juggling. We plan to introduce a new API function on the Curve class that will allow a bunch of keys to be moved in time atomically within the constraints of the undo system.

Reply:
Doh! Looks like I was writing my comment while you were writing yours...